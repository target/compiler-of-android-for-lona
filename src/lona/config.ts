import fs from 'fs'
import path from 'path'
import * as Options from '../options'

export type Config = {
  ignore: string[]
  android: Partial<Options.Raw>
}

export async function getConfig(workspacePath: string): Promise<Config> {
  const configPath = path.join(workspacePath, 'lona.json')

  const data = JSON.parse(await fs.promises.readFile(configPath, 'utf-8'))

  if (!data.ignore) {
    data.ignore = ['**/node_modules/**', '**/.git/**']
  }

  if (!data.android) {
    data.android = {}
  }

  if (data.android.output) {
    if (!data.android.output.startsWith('/')) {
      data.android.output = path.join(workspacePath, data.android.output)
    }
  }

  return data
}
