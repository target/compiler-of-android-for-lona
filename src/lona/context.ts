// import { Config } from "@lona/compiler/lib/utils/config"
// import { Reporter } from "@lona/compiler/lib/helpers/reporter"
// import path from 'path'
// import fs from 'fs'
// import * as LogicAST from "@lona/compiler/lib/helpers/logic-ast"
// import * as LogicScope from "@lona/compiler/lib/helpers/logic-scope"
// import * as LogicUnify from "@lona/compiler/lib/helpers/logic-unify"
// import * as LogicEvaluate from "@lona/compiler/lib/helpers/logic-evaluate"

// export const generate = (config: Config, reporter: Reporter) => {
//   const standardLibsPath = path.join(__dirname, '../../static/logic')
//   const standardLibs = fs.readdirSync(standardLibsPath)

//   const libraryFiles: LogicAST.AST.Program[] = standardLibs.map(
//     x =>
//       LogicAST.makeProgram(
//         JSON.parse(fs.readFileSync(path.join(standardLibsPath, x), 'utf8'))
//       ) as LogicAST.AST.Program
//   )

//   const standardLibsProgram = LogicAST.joinPrograms(libraryFiles)

//   const logicPrograms = Object.keys(config.logicFiles)
//     .map(k => {
//       const node = LogicAST.makeProgram(config.logicFiles[k])
//       if (!node) {
//         return undefined
//       }
//       return {
//         in: k,
//         node,
//       }
//     })
//     .filter(x => !!x)

//   const scopeContext = LogicScope.build(
//     [{ node: standardLibsProgram, in: 'standard library' }].concat(logicPrograms),
//     reporter
//   )

//   const programNode = LogicAST.joinPrograms([
//     standardLibsProgram,
//     ...logicPrograms.map(x => x.node),
//   ])

//   const unificationContext = LogicUnify.makeUnificationContext(
//     programNode,
//     scopeContext,
//     reporter
//   )
//   const substitution = LogicUnify.unify(
//     unificationContext.constraints,
//     reporter
//   )

//   const evaluationContext = LogicEvaluate.evaluate(
//     programNode,
//     programNode,
//     scopeContext,
//     unificationContext,
//     substitution,
//     reporter
//   )

//   return evaluationContext
// }
