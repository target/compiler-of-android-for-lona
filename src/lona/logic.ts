import * as Serialization from '@lona/serialization'
import * as FileSearch from '@lona/compiler/lib/utils/file-search'
import { copy, IFS } from 'buffs'
import path from 'path'

export function componentFiles(workspacePath: string): string[] {
  return FileSearch.sync(workspacePath, '**/*.cmp').map(file =>
    path.join(workspacePath, file)
  )
}

export function logicFiles(workspacePath: string): string[] {
  return FileSearch.sync(workspacePath, '**/*.logic').map(file =>
    path.join(workspacePath, file)
  )
}

export function documentFiles(workspacePath: string): string[] {
  return FileSearch.sync(workspacePath, '**/*.md').map(file =>
    path.join(workspacePath, file)
  )
}

export function libraryFiles(): string[] {
  const libraryPath = path.join(__dirname, 'library')

  return FileSearch.sync(libraryPath, '**/*.logic').map(file =>
    path.join(libraryPath, file)
  )
}

type FileTuple = [string, Serialization.LogicAST.SyntaxNode]

export function program(fs: IFS, workspacePath: string): any {
  function decode<T>(
    filePath: string,
    decoder: (data: string) => T
  ): [string, T] {
    const data = fs.readFileSync(filePath, 'utf8')
    const ast = decoder(data)
    return [filePath, ast]
  }

  const files: FileTuple[] = [
    ...libraryFiles().map(filePath =>
      decode(filePath, Serialization.decodeLogic)
    ),
    ...componentFiles(workspacePath).map(filePath =>
      decode(filePath, Serialization.decodeLogic)
    ),
    ...documentFiles(workspacePath).map(filePath =>
      decode(filePath, data =>
        Serialization.extractProgramFromAST(Serialization.decodeDocument(data))
      )
    ),
  ]

  console.log(files)
}
