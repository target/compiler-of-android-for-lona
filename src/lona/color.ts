import { ColorValue } from '@lona/compiler/lib/plugins/tokens/tokens-ast'
import * as XML from '../xml/ast'
import {
  createColor,
  cssToHexColor,
  formatQualifiedName,
  Options,
} from '../android/valueResources'

export type Token = {
  qualifiedName: string[]
  value: ColorValue
}

export const convert = (token: Token, options: Options): XML.Element => {
  return createColor(
    formatQualifiedName(token.qualifiedName, options),
    cssToHexColor(token.value.css)
  )
}
