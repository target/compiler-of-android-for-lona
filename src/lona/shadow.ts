import { ShadowValue } from '@lona/compiler/lib/plugins/tokens/tokens-ast'
import * as XML from '../xml/ast'
import {
  createDimen,
  formatQualifiedName,
  Options,
} from '../android/valueResources'

export type Token = {
  qualifiedName: string[]
  value: ShadowValue
}

export const convert = (token: Token, options: Options): XML.Element => {
  return createDimen(formatQualifiedName(token.qualifiedName, options), '2dp')
}
