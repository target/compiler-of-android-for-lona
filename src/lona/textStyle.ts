import { TextStyleValue } from '@lona/compiler/lib/plugins/tokens/tokens-ast'
import * as XML from '../xml/ast'
import { createStyle, createItem, cssToHexColor } from '../android/resources'

export type Token = {
  qualifiedName: string[]
  value: TextStyleValue
}

export type Options = {
  minSdk: number
}

export const convert = (token: Token, options: Options): XML.Element => {
  const {
    fontFamily,
    fontSize,
    fontWeight,
    color,
    lineHeight,
    letterSpacing,
  } = token.value

  let items: XML.Element[] = []

  if (fontFamily) {
    items.push(createItem('android:fontFamily', fontFamily))
  }

  if (fontSize) {
    items.push(createItem('android:textSize', `${fontSize}sp`))
  }

  if (options.minSdk >= 26 && fontWeight) {
    items.push(createItem('android:fontWeight', `${fontWeight}`))
  }

  if (color) {
    items.push(createItem('android:textColor', cssToHexColor(color.css)))
  }

  if (lineHeight && fontSize) {
    items.push(
      createItem('android:lineSpacingMultiplier', `${lineHeight / fontSize}`)
    )
  }

  if (letterSpacing) {
    items.push(createItem('android:letterSpacing', `${letterSpacing}`))
  }

  return createStyle(token.qualifiedName.join('_'), items)
}
