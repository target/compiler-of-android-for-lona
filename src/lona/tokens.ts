import { Token } from '@lona/compiler/lib/plugins/tokens/tokens-ast'

import * as XML from '../xml/ast'
import * as Color from './color'
import * as Shadow from './shadow'
import * as TextStyle from './textStyle'

function getColorTokens(tokens: Token[]): Color.Token[] {
  return tokens.flatMap(({ qualifiedName, value }) =>
    value.type === 'color'
      ? [{ qualifiedName: qualifiedName, value: value.value }]
      : []
  )
}

function getShadowsTokens(tokens: Token[]): Shadow.Token[] {
  return tokens.flatMap(({ qualifiedName, value }) =>
    value.type === 'shadow'
      ? [{ qualifiedName: qualifiedName, value: value.value }]
      : []
  )
}

function getTextStyleTokens(tokens: Token[]): TextStyle.Token[] {
  return tokens.flatMap(({ qualifiedName, value }) =>
    value.type === 'textStyle'
      ? [{ qualifiedName: qualifiedName, value: value.value }]
      : []
  )
}

export type TokenResources = {
  colors: XML.Element[]
  elevations: XML.Element[]
  textStyles: XML.Element[]
}

export function createValueResources(
  tokens: Token[],
  options: { minSdk: number; nameTemplate: string }
): TokenResources {
  const { nameTemplate } = options
  const valueResourceOptions = { nameTemplate }

  return {
    elevations: getShadowsTokens(tokens).map(token =>
      Shadow.convert(token, valueResourceOptions)
    ),
    colors: getColorTokens(tokens).map(token =>
      Color.convert(token, valueResourceOptions)
    ),
    textStyles: getTextStyleTokens(tokens).map(textStyle =>
      TextStyle.convert(textStyle, options, valueResourceOptions)
    ),
  }
}
