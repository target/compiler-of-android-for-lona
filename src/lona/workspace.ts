import { Helpers } from '@lona/compiler/lib/helpers'
import * as Tokens from '@lona/compiler/lib/plugins/tokens'
import { Token } from '@lona/compiler/lib/plugins/tokens/tokens-ast'
import * as FileSearch from '@lona/compiler/lib/utils/file-search'
import { copy, IFS } from 'buffs'
import fs from 'fs'
import path from 'path'
import { Union } from 'unionfs'
import util from 'util'
import {
  addDependenciesAndPluginsToModuleBuildScript,
  addKotlinDependencyToRootBuildScript,
} from '../android/buildScript'
import { formatDrawableName } from '../android/drawableResources'
import { createResourceFiles } from '../android/resources'
import { Validated } from '../options'
import { createFiles as createSvgDrawableFiles } from '../svg/drawable'
import { templatePathForName } from '../template/bundled'
import {
  createTemplateContext,
  CreateTemplateContextOptions,
} from '../template/context'
import { inflate, InflateOptions } from '../template/inflate'
import { createValueResources } from './tokens'

export function inflateProjectTemplate(
  outputPath: string,
  generateGallery: boolean,
  drawableResourceNames: string[],
  options: CreateTemplateContextOptions,
  inflateOptions?: InflateOptions
): { files: IFS; srcPath: string; resPath: string } {
  const { files: projectFiles } = inflate(
    fs,
    templatePathForName('project'),
    outputPath,
    createTemplateContext(options),
    inflateOptions
  )

  const fsAndProject = new Union()
  fsAndProject.use(fs)
  fsAndProject.use(projectFiles)

  const { files: moduleFiles, context: moduleContext } = inflate(
    fsAndProject,
    templatePathForName('module'),
    outputPath,
    createTemplateContext(options),
    inflateOptions
  )

  copy(moduleFiles, projectFiles, '/', '/')

  if (generateGallery) {
    const { files: galleryFiles, context: galleryContext } = inflate(
      fsAndProject,
      templatePathForName('gallery'),
      outputPath,
      createTemplateContext({
        ...options,
        overrides: {
          isNewProject: false,
          layoutName: 'activity_empty',
          activityClass: 'EmptyActivity',
          activityTitle: 'Empty Activity',
          isLauncher: false,
          relativePackage: undefined,
          drawableAssetList: drawableResourceNames,
        },
      }),
      inflateOptions
    )

    addKotlinDependencyToRootBuildScript(
      projectFiles,
      galleryContext,
      outputPath
    )
    addDependenciesAndPluginsToModuleBuildScript(
      projectFiles,
      galleryContext,
      outputPath
    )

    copy(galleryFiles, projectFiles, '/', '/')
  }

  return {
    files: projectFiles,
    srcPath: moduleContext.get('srcDir'),
    resPath: moduleContext.get('resDir'),
  }
}

async function convertTokens(workspacePath: string, helpers: Helpers) {
  const convertedWorkspace = await Tokens.parseWorkspace(
    workspacePath,
    helpers,
    {
      output: false,
    }
  )

  if (!convertedWorkspace) {
    return Promise.reject('Unknown problem while converting tokens')
  }

  const tokens = convertedWorkspace.files.flatMap(file =>
    file.contents.type === 'flatTokens' ? file.contents.value : []
  )

  return tokens
}

async function convertSvgFiles(
  workspacePath: string,
  nameTemplate: string,
  ignore: string[]
): Promise<[string, IFS][]> {
  const svgRelativePaths = FileSearch.sync(workspacePath, '**/*.svg', {
    ignore,
  })

  return Promise.all(
    svgRelativePaths.map(async relativePath => {
      const data = await fs.promises.readFile(
        path.join(workspacePath, relativePath)
      )

      const result: [string, IFS] = [
        relativePath,
        await createSvgDrawableFiles(relativePath, data, { nameTemplate }),
      ]

      return result
    })
  )
}

/**
 * Convert a Lona workspace to an Android library module.
 *
 * @returns A filesystem containing the generated library files.
 */
export async function convert(
  workspacePath: string,
  helpers: Helpers,
  options: Omit<Validated, 'shouldOutputFiles' | 'dryRun'>
): Promise<IFS> {
  const {
    verbose,
    outputPath,
    packageName,
    minSdk,
    buildSdk,
    targetSdk,
    generateGallery,
    valueResourceNameTemplate,
    drawableResourceNameTemplate,
  } = options

  if (verbose) {
    console.log('Converting with options:')
    console.log(util.inspect(options, false, null, true))
  }

  let tokens: Token[]

  try {
    tokens = await convertTokens(workspacePath, helpers)
  } catch (error) {
    helpers.reporter.error('Failed to convert tokens')
    return Promise.reject(error)
  }

  const drawableResources: [string, IFS][] = await convertSvgFiles(
    workspacePath,
    valueResourceNameTemplate,
    helpers.config.ignore
  )

  const drawableResourceNames = drawableResources.map(([key]) =>
    formatDrawableName(key, undefined, {
      nameTemplate: drawableResourceNameTemplate,
    })
  )

  const { files: templateFiles, resPath } = inflateProjectTemplate(
    outputPath,
    generateGallery,
    drawableResourceNames,
    { packageName, minSdk, buildSdk, targetSdk },
    { verbose }
  )

  const valueResources = createValueResources(tokens, {
    minSdk,
    nameTemplate: drawableResourceNameTemplate,
  })

  const { fs: resourceFiles } = createResourceFiles(
    path.join(outputPath, resPath),
    {
      colorResources: valueResources.colors,
      elevationResources: valueResources.elevations,
      textStyleResources: valueResources.textStyles,
      drawableResources,
    }
  )

  copy(resourceFiles, templateFiles, '/', '/')

  return templateFiles
}
