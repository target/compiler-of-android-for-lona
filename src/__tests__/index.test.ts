import path from 'path'
import { describe as describeFs } from 'buffs'
import { convert } from '../lona/workspace'
import createHelpers from '@lona/compiler/lib/helpers'

describe('Convert', () => {
  it('converts the example workspace', async () => {
    const workspacePath = path.join(__dirname, '../../example')
    const outputPath = '/build'

    const helpers = await createHelpers(workspacePath)
    const result = await convert(workspacePath, helpers, {
      verbose: false,
      outputPath,
      packageName: 'com.test.designlibrary',
      minSdk: 21,
      buildSdk: 29,
      targetSdk: 29,
      generateGallery: false,
    })

    expect(describeFs(result, outputPath)).toMatchSnapshot()
  })
})
