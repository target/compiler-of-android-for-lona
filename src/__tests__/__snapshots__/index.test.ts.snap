// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Convert converts the example workspace 1`] = `
"build
├── .gitignore
├── androidTest / com / test / designlibrary / ApplicationTest.java
├── build.gradle
├── designlibrary
│   ├── .gitignore
│   ├── build.gradle
│   ├── libs
│   └── src
│       ├── com / test / designlibrary
│       └── main
│           ├── AndroidManifest.xml
│           └── res
│               ├── drawable / assets_check_circle.xml
│               ├── drawable-hdpi / assets_gradient_fill.png
│               ├── drawable-mdpi / assets_gradient_fill.png
│               ├── drawable-xhdpi / assets_gradient_fill.png
│               ├── drawable-xxhdpi / assets_gradient_fill.png
│               ├── drawable-xxxhdpi / assets_gradient_fill.png
│               └── values
│                   ├── colors.xml
│                   ├── elevations.xml
│                   ├── strings.xml
│                   └── text-styles.xml
├── gradle / wrapper
│   ├── gradle-wrapper.jar
│   └── gradle-wrapper.properties
├── gradle.properties
├── gradlew
├── gradlew.bat
├── local.properties
└── settings.gradle"
`;
